src/osd/ReplicatedBackend.cc

```diff
@@ -1755,11 +1817,16 @@ bool ReplicatedBackend::handle_pull_response(
   dout(10) << "new recovery_info " << pi.recovery_info
 	   << ", new progress " << pi.recovery_progress
 	   << dendl;
-
+  interval_set<uint64_t> data_zeros;
+  uint64_t z_offset = pop.before_progress.data_recovered_to;
+  uint64_t z_length = pop.after_progress.data_recovered_to - pop.before_progress.data_recovered_to;
+  if(z_length)
+    data_zeros.insert(z_offset, z_length);
   bool complete = pi.is_complete();
-
+  bool clear_omap = !pop.before_progress.omap_complete;
```

Comment:
omap recovery can span more than one push op now, so we should only clear omap for the first one